src/python/pants/engine/process.py

```diff
@@ -376,12 +402,18 @@ async def find_binary(request: BinaryPathRequest) -> BinaryPaths:
     search_path = create_path_env_var(request.search_path)
     result = await Get(
         FallibleProcessResult,
-        Process(
-            description=f"Searching for `{request.binary_name}` on PATH={search_path}",
-            level=LogLevel.DEBUG,
-            input_digest=script_digest,
-            argv=[script_path, request.binary_name],
-            env={"PATH": search_path},
+        # We use a volatile process to force re-run since any binary found on the host system today
+        # could be gone tomorrow. Ideally we'd only do this for local processes since all known
+        # remoting configurations include a static container image as part of their cache key which
+        # automatically avoids this problem.
+        VolatileProcess(
```

Comment:
Aaaaand ... this is _still_ broken in realistic ways. Namely, upgrading or downgrading a binary found here previously will result in the same output from this rule and thus trigger no re-run of rules depending on this data, which is wrong. The only cases currently covered by this fix from the list below are 1 and 2:
1. An applicable binary is added to the search path.
2. An applicable binary is removed from the search path.
3. An applicable binary on the search path is modified (upgraded or downgraded most likely).

To handle 3 the contents of the discovered binaries need to be hashed. This could be done by storing it in the CAS as a hack, but we won't / can't generally use the binary stored in the CAS so that just adds more (see `make_process_volatile` comment) storage bloat. Ideally we'd ask an engine intrinsic to get us a fingerprint without storing the object. It could also be done by just hashing the resulting absolute path in python rule code at the expense of potentially heavy IO inhibiting parallelism. On my machine, hashing my biggest binary (/usr/bin/packer @ 154M) takes ~400ms in pure python code and hashing the currently more typical python binary (@ 19M) takes ~50ms.

I'd like to defer handling case 3 to a follow-up if folks are amenable.

Reply:
Indeed. Tom had suggested hashing the binaries, as well. My original design was going to update the `find_binary` script to also run `shasum -a 256` if available, then to store a private field `_hashes` on the `BinaryPaths` object. Call sites wouldn't use this, but it would result in the cache becoming invalidated if the hashes change, even if the paths remain the same.

I don't think we need to block on #3. By definition, fixing #1 and #2 will be an improvement from what we have right now. 

We should track this, though. It could result in really subtle issues for users that require them knowing to purge `lmdb_store`..

--

It gets even worse for Pyenv, which was the whole motivation of this fix. `~/.pyenv/shims/python` is a Bash script. It will always have the same hash, regardless of what interpreter is used. The only "fixes" I could imagine are a) calling `pyenv which python` somehow, or b) telling users to stop using Pyenv how you're supposed to, e.g. don't use `pyenv global` and instead explicitly list the `.pyenv/versions` on their PATH.

Reply:
Yeah, I think shasum if available is not a great idea. Even relying on bash and optional which as is is bad and TODO'd for replacement with a native binary for each platform we support.

Pyenv shims would be solved IIUC by something this should already be doing, which is allowing a test function. Really a test argv where each found binary is executed as `argv0 + argvTest`. After all, the found binary is just name matched. It may be broken, it may be malicious, etc. A test function at least provides a nominal check the binary works and the result of the output of test function, if any, could be mixed into the hash.

Reply:
I'll follow up with this change and temporarily push hashing into the check process with an issue broken out to separate this out. Since python is an interpreter we only happen to be able to do fancy checks with it - for other binaries that won't be possible.

Reply:
The follow-up is in #10770.